clients/typescript/src/sockets/web.ts

```diff
@@ -2,77 +2,124 @@ import { ConnectionOptions, Data, Socket, SocketFactory } from '.'
 import { SatelliteError, SatelliteErrorCode } from '../util'
 
 export class WebSocketWebFactory implements SocketFactory {
-  create() {
+  create(): WebSocketWeb {
     return new WebSocketWeb()
   }
 }
 
+// FIXME: This implementation is a bit contrived because it is not using EventEmitter
 export class WebSocketWeb implements Socket {
   private socket?: WebSocket
 
-  private connectCallbacks: (() => void)[]
-  private errorCallbacks: ((error: Error) => void)[]
+  private connectCallbacks: (() => void)[] = []
+  private errorCallbacks: ((error: SatelliteError) => void)[] = []
+  private onceErrorCallbacks: ((error: SatelliteError) => void)[] = []
 
-  constructor() {
-    this.connectCallbacks = []
-    this.errorCallbacks = []
+  // event doesn't provide much
+  private errorListener = () => {
+    for (const cb of this.errorCallbacks) {
+      cb(new SatelliteError(SatelliteErrorCode.SOCKET_ERROR, 'socket error'))
+    }
+
+    while (this.onceErrorCallbacks.length > 0) {
+      const cb = this.onceErrorCallbacks.pop()
+      if (cb)
+        cb(new SatelliteError(SatelliteErrorCode.SOCKET_ERROR, 'socket error'))
+    }
+  }
+
+  private connectListener = () => {
+    while (this.connectCallbacks.length > 0) {
+      this.connectCallbacks.pop()!()
+    }
```

Comment:
This can be written slightly more error-resistant without the `!`
```suggestion
    let callback: (() => void) | undefined
    while (callback = this.connectCallbacks.pop()) {
      callback()
    }
```